osu.Game/Audio/BassAudioNormalization.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using ManagedBass;
+using ManagedBass.Loud;
+using osu.Framework.Logging;
+
+namespace osu.Game.Audio
+{
+    /// <summary>
+    /// Audio Normalization Implementation using Bass
+    /// </summary>
+    public class BassAudioNormalization
+    {
+        /// <summary>
+        /// The integrated loudness of the audio
+        /// </summary>
+        /// <remarks>
+        /// Applicable range = -70 to -1<br/>Null if the loudness could not be calculated due to an error
+        /// </remarks>
+        public float? IntegratedLoudness { get; }
+
+        /// <summary>
+        /// Calculate the integrated loudness of an audio file using Bass
+        /// </summary>
+        /// <returns>The integrated loudness or 1 in <see cref="IntegratedLoudness"/></returns>
+        /// <param name="filePath">A path to an audio file</param>
+        public BassAudioNormalization(string filePath)
+        {
+            int decodeStream = Bass.CreateStream(filePath, 0, 0, BassFlags.Decode | BassFlags.Float);
+
+            if (decodeStream == 0)
+            {
+                Logger.Log("Failed to create stream for loudness measurement!\nError Code: " + Bass.LastError, LoggingTarget.Runtime, LogLevel.Error);
+                IntegratedLoudness = null;
+                Bass.StreamFree(decodeStream);
+                return;
+            }
+
+            int loudness = BassLoud.BASS_Loudness_Start(decodeStream, BassFlags.BassLoudnessIntegrated | BassFlags.BassLoudnessAutofree, 0);
+
+            if (loudness == 0)
+            {
+                Logger.Log("Failed to start loudness measurement!\nError Code: " + Bass.LastError, LoggingTarget.Runtime, LogLevel.Error);
+                IntegratedLoudness = null;
+                Bass.StreamFree(decodeStream);
+                return;
+            }
+
+            byte[] buffer = new byte[10000];
+
+            while (Bass.ChannelGetData(decodeStream, buffer, buffer.Length) >= 0)
+            {
+            }
```

Comment:
I'm not sure if this is correct.

The C code that Ian uses as an example seems to create a buffer with the size of 20000 on every run of the while loop:
```c
while (1) {
	char buffer[20000];
	int c = BASS_ChannelGetData(decoder, buffer, sizeof(buffer));
	if (c < 0) break;
}
```
While here we create a single buffer with the size of 10000, then run the while loop so long as ChannelGetData is greater than or equal to 0. I'm slightly concerned that this approach might not actually be processing the entire file since we have the buffer creation outside of the while loop.

Reply:
> as a compiler will generally optimize it to create the buffer only once

Will it? I'm not that certain it will. Linking cppreference doesn't exactly inspire confidence in the veracity of this claim.

If you're going to claim this actually look into the CIL or JITted code via sharplab or something, or profile using actual C# tools.

Reply:
> as a compiler will generally optimize it to create the buffer only once

That comment is on the C example, not on any C# code.

Reply:
...but you're making a claim that the compiler will do things in C# code based on a C example? i don't get it?

please test against the actual things in verifiable and reproducible ways rather than extrapolate.

Reply:
Post edited to clarify further. Nothing is really implied about C# code.

Reply:
I guess I'll leave this as it looks, but I still want to change the buffer size to be 20000 just to be consistent with Ian's example.